numpy/core/include/numpy/ndarraytypes.h

```diff
@@ -1027,6 +1027,8 @@ typedef void (NpyIter_GetMultiIndexFunc)(NpyIter *iter,
  * eliminate overlap.
  */
 #define NPY_ITER_COPY_IF_OVERLAP            0x00002000
+/* At least one operand uses UPDATEIFCOPY semantics, use context manager */
+#define NPY_ITER_USE_CONTEXT_MANAGER        0x00004000
```

Comment:
Simply resolving the ``UPDATEIFCOPY`` semantics on __exit__ mean that the iterator is still, in some sense, valid where calling ``NpyIter_Deallocate`` would totally destroy any internal state of the iterator.

Calling ``NpyIter_Deallocate`` should trigger calls to ``array_dealloc`` for each ``PyArrayObject`` operand, and ``array_dealloc`` calls ``PyArray_ResolveUpdateIfCopy``. So any of the internal calls to ``NpyIter_*iNew`` inside numpy, which all are terminated by a call to ``NpyIter_Deallocate``, are safe from the ``UPDATEIFCOPY`` problem, which means ``NpyIter`` should also be safe. It is only the app-level ``numpy.nditer`` that must be protected with a context manager.

However I could imagine a scenario where somehow during an iteration, an app-level subclass of ndarray calls an app-level function, which then causes PyPy to create an internal (app-level accessable) representation of the ndarray operand, whose reference will now be managed by the PyPy garbage collector rather than by pure C level refcount semantics. In this case[ ``Py_XDECREF(*object)``](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/nditer_constr.c#L694) will not trigger ``array_dealloc``, putting us back to square one.

I am pretty sure this cannot ever happen (famous last words)